design/proposals/module-linking/Explainer.md

```diff
@@ -838,34 +842,60 @@ binaries in all the same places where core module binaries can be loaded today,
+write, in JavaScript:
+```js
+import Foo from "./foo.wasm" as "wasm-module";
+assert(Foo instanceof WebAssembly.Module);
+```
+With this extension to JavaScript and the ESM loader, an adapter module import
+of module type can be treated like an `as "wasm-module"` import by
+ESM-integration.
 
-For single-level imports of non-instance type, the natural analogue in an ESM
-context is the [default export]. Thus, a single-level non-instance import would
-receive whatever the ESM loader says the default value would be.
+In all other cases, the (single) string imported by an adapter module import is
+first resolved to a [Module Record] using the same process as resolving the
+[Module Specifier] of a JavaScript `import`. After this, the handling of the
+imported Module Record is determined by the import type:
 
 For imports of instance type, the ESM loader would treat the exports of the
-instance type as the imported field names. In the normal two-level
-import-an-instance case, this corresponds exactly to existing ESM-integration
-behavior. However, for more-deeply nested instances, the ESM loader would need
-to recursively extract fields.
-
-Lastly, for imports of modules, ESM-integration would need the ESM loader to
-provide a fundamentally new way to parse/decode a module without *also*
-instantiating that module. Since this functionality is also useful for JS,
-there is already a proposal to add this to ESM called [Import Reflection].
-With this proposal, a module import:
+instance type as if they were the [Named Imports] of a JavaScript `import`.
+Thus, single-level imports of instance type mirror the behavior of core
+two-level imports with existing ESM-integration. Since the exports of an
+instance type can themselves be instance types, this process must be performed
+recursively.
+
+For imports of type `(global externref)` (i.e., an immutable core WebAssembly
+global storing a value of type `externref`), the imported Module Record is
+first converted to a [Namespace Object] via [`GetModuleNamespace()`] and then
+converted to a `(global externref)` as a normal JavaScript value. This allows
+the importing adapter module to dynamically access the fields of the
+Namespace Object by calling other imported functions (such as `Object.get()`).
+In the future, other compatible types (e.g., `handle` Interface Types) could be
+included in this case.
+
+Otherwise, if the import cannot accept a Namespace Object (and would error
+otherwise), the import is treated like a JavaScript [Imported Default Binding]
+and the Module Record is converted to its default value. This allows, for
+example, a single level import of a function:
 ```wasm
 (adapter module
-  (import "./foo.wasm" (module $Foo))
+  (import "./foo.js" (func (result i32)))
 )
 ```
-could be handled by ESM-integration as-if loaded by:
+to be satisfied by a JavaScript module via ESM-integration:
 ```js
-import Foo from "./foo.wasm" as "wasm-module";
+// foo.js
+export default () => 42;
```

Comment:
While sugaring the `"default"` import definitely makes sense, it does preclude the ability to obtain other function export names from JS or Wasm. Does this mean an adapter module cannot import a function or value in the ESM integration from a JS module without a default export?

Reply:
I might be missing the corner case you're tickling here, but to import a function or value using ESM integration from a JS module without a default export, you'd use an instance import, in which case you're not hitting the default sugar fallback case.  Or do you mean something else?

Reply:
Ok to clarify I'm understanding, are you saying that a JS module is treated as an instance just like a Wasm module is? If so it seems I overlooked that!

Reply:
Yep!  I believe the chain of reasoning here is: a JS module is loaded via the ESM loader to produce a Module Record, and the ESM-integration path is able to directly satisfy an instance-typed import with a Module Record.